File: src/main/java/org/voegtle/weatherwidget/WeatherActivity.java
package org.voegtle.weatherwidget;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import org.voegtle.weatherwidget.base.ThemedActivity;
import org.voegtle.weatherwidget.diagram.*;
import org.voegtle.weatherwidget.location.LocationIdentifier;
import org.voegtle.weatherwidget.location.LocationView;
import org.voegtle.weatherwidget.location.WeatherLocation;
import org.voegtle.weatherwidget.preferences.ApplicationSettings;
import org.voegtle.weatherwidget.preferences.ColorScheme;
import org.voegtle.weatherwidget.preferences.WeatherPreferences;
import org.voegtle.weatherwidget.preferences.WeatherSettingsReader;
import org.voegtle.weatherwidget.util.StatisticsUpdater;
import org.voegtle.weatherwidget.util.WeatherDataUpdater;

import java.util.HashMap;

public class WeatherActivity extends ThemedActivity implements SharedPreferences.OnSharedPreferenceChangeListener {
  private WeatherDataUpdater updater;
  private StatisticsUpdater statisticsUpdater;
  private ApplicationSettings configuration;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);

    setContentView(R.layout.activity_weather);
    initButton(R.id.button_google_docs, Uri.parse("https://docs.google.com/spreadsheets/d/1ahkm9SDTqjYcsLgKIH5yjmqlAh6dKxgfIrZA5Dt9L3o/edit?usp=sharing"));

    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
    readConfiguration(preferences);
    preferences.registerOnSharedPreferenceChangeListener(this);

    statisticsUpdater = new StatisticsUpdater(this);

    setupLocations();
    configureLocationSymbolColor();

    updater = new WeatherDataUpdater(this, configuration);
  }

  private void startWeatherUpdater() {
    updater.startWeatherScheduler(180);
  }

  private void setupLocations() {
    for (WeatherLocation location : configuration.getLocations()) {
      addClickHandler(location);
      updateVisibility(location);
      updateState(location);
      updateTextSize(location, configuration.getAppTextSize());
    }
  }

  private void readConfiguration(SharedPreferences preferences) {
    WeatherSettingsReader weatherSettingsReader = new WeatherSettingsReader(this.getApplicationContext());
    configuration = weatherSettingsReader.read(preferences);
  }

  private void addClickHandler(final WeatherLocation location) {
    final LocationView locationView = (LocationView) findViewById(location.getWeatherViewId());
    locationView.setOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        if (locationView.isExpanded()) {
          statisticsUpdater.updateStatistics(locationView, location);
        } else {
          statisticsUpdater.clearState(locationView);
        }
      }
    });

    locationView.setDiagramsOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        switch (view.getId()) {
          case R.id.weather_paderborn:
            startActivity(new Intent(WeatherActivity.this, PaderbornDiagramActivity.class));
            break;
          case R.id.weather_bali:
            startActivity(new Intent(WeatherActivity.this, BaliDiagramActivity.class));
            break;
          case R.id.weather_bonn:
            startActivity(new Intent(WeatherActivity.this, BonnDiagramActivity.class));
            break;
          case R.id.weather_freiburg:
            startActivity(new Intent(WeatherActivity.this, FreiburgDiagramActivity.class));
            break;
          case R.id.weather_leo:
            startActivity(new Intent(WeatherActivity.this, LeoDiagramActivity.class));
            break;
          case R.id.weather_herzogenaurach:
            startActivity(new Intent(WeatherActivity.this, HerzoDiagramActivity.class));
            break;
          case R.id.weather_mobil:
            Intent intent = new Intent(WeatherActivity.this, MobilDiagramActivity.class);
            WeatherLocation location = configuration.findLocation(LocationIdentifier.Mobil);
            if (location != null) {
              intent.putExtra(MobilDiagramActivity.class.getName(), location.getName());
            }
            startActivity(intent);
            break;
        }
      }
    });

    locationView.setForecastOnClickListener(new View.OnClickListener() {
      @Override
      public void onClick(View view) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, location.getForecastUrl());
        startActivity(browserIntent);
      }
    });

  }

  private void updateStatistics() {
    HashMap<LocationView, WeatherLocation> updateCandidates = new HashMap<>();
    for (WeatherLocation location : configuration.getLocations()) {
      final LocationView locationView = (LocationView) findViewById(location.getWeatherViewId());
      if (locationView.isExpanded()) {
        updateCandidates.put(locationView, location);
      }
    }
    statisticsUpdater.updateStatistics(updateCandidates, false);
  }

  private void updateVisibility(WeatherLocation location) {
    boolean show = location.getPreferences().isShowInApp();
    updateVisibility(location.getWeatherViewId(), show);
  }

  private void updateVisibility(int viewId, boolean isVisible) {
    View view = findViewById(viewId);
    view.setVisibility(isVisible ? View.VISIBLE : View.GONE);
  }

  private void updateTextSize(WeatherLocation location, Integer textSize) {
    if (location.getPreferences().isShowInApp()) {
      LocationView view = (LocationView) findViewById(location.getWeatherViewId());
      view.setTextSize(textSize);
    }
  }

  private void configureLocationSymbolColor() {
    // Dunkle Symbole wenn der Hintergrund hell ist
    boolean darkSymbols = getColorScheme() == ColorScheme.light;
    for (WeatherLocation location : configuration.getLocations()) {
      final LocationView locationView = (LocationView) findViewById(location.getWeatherViewId());
      locationView.configureSymbols(darkSymbols);
    }

  }

  private void updateState(WeatherLocation location) {
    final LocationView locationView = (LocationView) findViewById(location.getWeatherViewId());
    statisticsUpdater.setupStatistics(locationView);
  }


  private void initButton(int buttonId, final Uri uri) {
    Button button = (Button) findViewById(buttonId);

    button.setOnClickListener(new View.OnClickListener() {

      @Override
      public void onClick(View arg0) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, uri);
        startActivity(browserIntent);
      }

    });
  }

  @Override
  protected void onResume() {
    super.onResume();
    updateStatistics();
    updater.updateWeatherOnce(false);
    startWeatherUpdater();
  }

  @Override
  protected void onPause() {
    super.onPause();
    updater.stopWeatherScheduler();
  }

  @Override
  public boolean onCreateOptionsMenu(Menu menu) {
    MenuInflater inflater = getMenuInflater();
    inflater.inflate(R.menu.weather_activity_menu, menu);
    return super.onCreateOptionsMenu(menu);

  }

  @Override
  public boolean onOptionsItemSelected(MenuItem item) {
    switch (item.getItemId()) {
      case R.id.action_reload:
        updater.updateWeatherOnce(true);
        return true;
      case R.id.action_diagrams:
        startActivity(new Intent(this, MainDiagramActivity.class));
        return true;
      case R.id.action_perferences:
        startActivity(new Intent(this, WeatherPreferences.class));
        return true;
    }
    return false;
  }

  @Override
  public void onSharedPreferenceChanged(SharedPreferences preferences, String s) {
    readConfiguration(preferences);
    setupLocations();
    updater.stopWeatherScheduler();
    updater = new WeatherDataUpdater(this, configuration);
  }
}

File: src/main/java/org/voegtle/weatherwidget/WeatherActivity.kt
package org.voegtle.weatherwidget

import android.Manifest
import android.app.AlertDialog
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.preference.PreferenceManager
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.view.Menu
import android.view.MenuItem
import android.view.View
import kotlinx.android.synthetic.main.activity_weather.*
import org.voegtle.weatherwidget.base.ThemedActivity
import org.voegtle.weatherwidget.base.UpdatingScrollView
import org.voegtle.weatherwidget.diagram.*
import org.voegtle.weatherwidget.location.LocationIdentifier
import org.voegtle.weatherwidget.location.LocationOrderStore
import org.voegtle.weatherwidget.location.LocationView
import org.voegtle.weatherwidget.location.WeatherLocation
import org.voegtle.weatherwidget.preferences.*
import org.voegtle.weatherwidget.util.ActivityUpdateTask
import org.voegtle.weatherwidget.util.StatisticsUpdater
import org.voegtle.weatherwidget.util.UserFeedback
import java.util.*


class WeatherActivity : ThemedActivity(), SharedPreferences.OnSharedPreferenceChangeListener {
  companion object {
    val ANDROID8 = "org.voegtle.weatherwidget.Android8"
  }

  private var statisticsUpdater: StatisticsUpdater? = null
  private var configuration: ApplicationSettings? = null

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)

    setContentView(R.layout.activity_weather)
    button_google_docs.setOnClickListener {
      val browserIntent = Intent(Intent.ACTION_VIEW,
                                 Uri.parse(
                                     "https://docs.google.com/spreadsheets/d/1ahkm9SDTqjYcsLgKIH5yjmqlAh6dKxgfIrZA5Dt9L3o/edit?usp=sharing"))
      startActivity(browserIntent)
    }

    val preferences = PreferenceManager.getDefaultSharedPreferences(this)
    readConfiguration(preferences)
    preferences.registerOnSharedPreferenceChangeListener(this)

    statisticsUpdater = StatisticsUpdater(this)

    setupLocations()
    configureLocationSymbolColor()


    scroll_view.register(object : UpdatingScrollView.Updater {
      override fun update() {
        updateWeatherOnce(true)
      }
    })

    if (intent.action == ANDROID8) {
      showAndroid8Explanation()
    }
  }

  private fun showAndroid8Explanation() {
    val builder = AlertDialog.Builder(this)
    builder.setMessage(R.string.android8_explanation).setTitle(R.string.android8_caption)
    val dialog = builder.create()
    dialog.show()
  }

  private fun setupLocations() {
    configuration?.let {
      it.locations.forEach { location ->
        addClickHandler(location)
        updateVisibility(location)
        updateState(location)
        updateTextSize(location, it.appTextSize)
      }
    }
  }

  private fun readConfiguration(preferences: SharedPreferences) {
    val weatherSettingsReader = WeatherSettingsReader(this.applicationContext)
    configuration = weatherSettingsReader.read(preferences)
    requestLocationPermission()
  }

  private fun addClickHandler(location: WeatherLocation) {
    val locationView: LocationView = findViewById(location.weatherViewId)
    locationView.setOnClickListener {
      if (locationView.isExpanded) {
        statisticsUpdater!!.updateStatistics(locationView, location)
      } else {
        statisticsUpdater!!.clearState(locationView)
      }
    }

    locationView.diagramListener = View.OnClickListener { view ->
      when (view.id) {
        R.id.weather_paderborn -> startActivity(Intent(this@WeatherActivity, PaderbornDiagramActivity::class.java))
        R.id.weather_bali -> startActivity(Intent(this@WeatherActivity, BaliDiagramActivity::class.java))
        R.id.weather_bonn -> startActivity(Intent(this@WeatherActivity, BonnDiagramActivity::class.java))
        R.id.weather_freiburg -> startActivity(Intent(this@WeatherActivity, FreiburgDiagramActivity::class.java))
        R.id.weather_leo -> startActivity(Intent(this@WeatherActivity, LeoDiagramActivity::class.java))
        R.id.weather_herzogenaurach -> startActivity(Intent(this@WeatherActivity, HerzoDiagramActivity::class.java))
        R.id.weather_magdeburg -> startActivity(Intent(this@WeatherActivity, MagdeburgDiagramActivity::class.java))
        R.id.weather_shenzhen -> startActivity(Intent(this@WeatherActivity, ShenzhenDiagramActivity::class.java))
        R.id.weather_mobil -> {
          val intent = Intent(this@WeatherActivity, MobilDiagramActivity::class.java)
          val mobileLocation = configuration!!.findLocation(LocationIdentifier.Mobil)
          if (mobileLocation != null) {
            intent.putExtra(MobilDiagramActivity::class.java.name, mobileLocation.name)
          }
          startActivity(intent)
        }
      }
    }

    locationView.forecastListener = View.OnClickListener {
      val browserIntent = Intent(Intent.ACTION_VIEW, location.forecastUrl)
      startActivity(browserIntent)
    }
  }

  private fun updateStatistics() {
    val updateCandidates = HashMap<LocationView, WeatherLocation>()
    configuration?.let {
      it.locations.forEach { location ->
        val locationView: LocationView = findViewById(location.weatherViewId)
        if (locationView.isExpanded) {
          updateCandidates.put(locationView, location)
        }
      }
    }
    statisticsUpdater?.updateStatistics(updateCandidates, false)
  }

  private fun updateVisibility(location: WeatherLocation) {
    val show = location.preferences.showInApp
    updateVisibility(location.weatherViewId, show)
  }

  private fun updateVisibility(viewId: Int, isVisible: Boolean) {
    val view: View = findViewById(viewId)
    view.visibility = if (isVisible) View.VISIBLE else View.GONE
  }

  private fun updateTextSize(location: WeatherLocation, textSize: Int) {
    if (location.preferences.showInApp) {
      val view: LocationView = findViewById(location.weatherViewId)
      view.setTextSize(textSize)
    }
  }

  private fun configureLocationSymbolColor() {
    // Dunkle Symbole wenn der Hintergrund hell ist
    val darkSymbols = colorScheme == ColorScheme.light
    configuration?.let {
      it.locations.forEach {
        val view: LocationView = findViewById(it.weatherViewId)
        view.configureSymbols(darkSymbols)
      }
    }

  }

  private fun updateState(location: WeatherLocation) {
    val locationView: LocationView = findViewById(location.weatherViewId)
    statisticsUpdater!!.setupStatistics(locationView)
  }

  override fun onResume() {
    super.onResume()
    updateStatistics()
    updateWeatherOnce(false)
  }

  override fun onPause() {
    super.onPause()
  }

  override fun onCreateOptionsMenu(menu: Menu): Boolean {
    val inflater = menuInflater
    inflater.inflate(R.menu.weather_activity_menu, menu)
    return super.onCreateOptionsMenu(menu)
  }

  override fun onOptionsItemSelected(item: MenuItem): Boolean =
      when (item.itemId) {
        R.id.action_reload -> {
          updateWeatherOnce(true)
          true
        }
        R.id.action_diagrams -> {
          startActivity(Intent(this, MainDiagramActivity::class.java))
          true
        }
        R.id.action_sort -> {
          val orderCriteriaDialog = OrderCriteriaDialogBuilder.createOrderCriteriaDialog(this)
          orderCriteriaDialog.show()
          true
        }
        R.id.action_perferences -> {
          startActivity(Intent(this, WeatherPreferences::class.java))
          true
        }
        else -> false
      }

  override fun onSharedPreferenceChanged(preferences: SharedPreferences, s: String) {
    readConfiguration(preferences)
    setupLocations()
    updateWeatherOnce(true)
  }

  fun requestLocationPermission() {
    val locationOrderStore = LocationOrderStore(applicationContext)
    val orderCriteria = locationOrderStore.readOrderCriteria()
    if (orderCriteria == OrderCriteria.location) {
      val permissionCheck = ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
      if (permissionCheck != PackageManager.PERMISSION_GRANTED) {
        ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION), 0)
      }
    }
  }

  override fun onRequestPermissionsResult(requestId: Int, permissions: Array<String>, grantResults: IntArray) {
    if (grantResults.isNotEmpty()) {
      if (grantResults[0] != PackageManager.PERMISSION_GRANTED) {
        UserFeedback(this).showMessage(R.string.message_location_permission_required, true)
        val locationOrderStore = LocationOrderStore(applicationContext)
        locationOrderStore.writeOrderCriteria(OrderCriteria.default)
        updateWeatherOnce(false)
      }
    }
  }

  fun updateWeatherOnce(showToast: Boolean) {
    ActivityUpdateTask(this, configuration!!, showToast).execute()
  }


}
